plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py

```diff
@@ -0,0 +1,213 @@
+            del configuration['device']
+            configuration["zpool"] = pool_name
+
+        elif 'vdev' in configuration:
+            vdev_defn = configuration['vdev'].split(' ')
+            # check no word attempts to play tricks us passing arbitrary options
+            for word in vdev_defn:
+                if not (word[0].isalpha() or word[0] == "/"):
+                    raise Exception('"vdev" contain invalid-looking string %r' % (word,))
+
+            pool_name = "sr-{}".format(sr_uuid)
+            zfsutils.pool_create(dbg, pool_name, vdev_defn)
+
+            # "vdev" is only used once to create the zpool, which
+            # then becomes the sole way to designate the SR
+            configuration["orig-vdev"] = configuration['vdev']
+            del configuration['vdev']
+            configuration["zpool"] = pool_name
+
+        else:
+            log.error('devices config must have "zpool", "vdev", or "device"')
+            raise Exception('devices config must have "zpool", "vdev", or "device"')
+
+        # FIXME this assumes zpool is mounted/attached
+        mountpoint = zfsutils.pool_mountpoint(dbg, pool_name)
+        importlib.import_module('zfs-vol').Callbacks().create_database(mountpoint)
+
+        meta = {
+            # mandatory elements we need everywhere
+            'name': name,
+            'description': description,
+            'uuid': sr_uuid,
+            # pool name may not always be derived from mountpoint or
+            # sr_uuid, esp. when creating with "zpool=$PREBUILT_POOL"
+            'zpool': pool_name,
+        }
+        util.update_sr_metadata(dbg, 'file://' + mountpoint, meta)
+
+        log.debug('{}: SR.create: sr={}'.format(dbg, mountpoint))
+        return configuration
+
+    def destroy(self, dbg, sr):
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        zfsutils.pool_destroy(dbg, meta["zpool"])
+
+    def attach(self, dbg, configuration):
+        log.debug('{}: SR.attach: config={}'.format(dbg, configuration))
+
+        # ZFS automagically attaches a pool to a mountpoint on
+        # create/boot/etc, so we basically do nothing here but find
+        # this mountpoint
+
+        # FIXME: study how pools are mounted on boot and see if we
+        # could do real a/detach
+
+        return zfsutils.pool_mountpoint(dbg, configuration["zpool"])
+
+    def detach(self, dbg, sr):
+        # Nothing to unmount for now.
+        pass
+
+    def ls(self, dbg, sr):
+        results = []
+        meta = util.get_sr_metadata(dbg, 'file://' + sr)
+        pool_name = meta["zpool"]
+        cb = importlib.import_module('zfs-vol').Callbacks()
+        with VolumeContext(cb, sr, 'r') as opq:
+            with cb.db_context(opq) as db:
+                vdis = db.get_all_vdis()
+                all_custom_keys = db.get_all_vdi_custom_keys()
+
+            for vdi in vdis:
+                # TODO: handle this better
+                # _vdi_sanitize(vdi, opq, db, cb)
```

Comment:
Shouldn't we call it instead of commenting? If we need to handle this better maybe it should be done in another patch that will modify `_vdi_sanitize`?

Reply:
It's not that easy as this function is not applicable, but something similar is part of last weeks work still in fixup commits, see ab2dc908966e6724ab8179eeed5271113314d245 and later

Reply:
My bad, that one being in `sr.py` not `volume.py` I missed it.  Addressed in upcoming push.